### PR TITLE
feat: [Do Not Mrege] Add Sparse Float Vector support to milvus

### PR DESCRIPTION
**Pull request by zhengbuqian**
issue: #29419 

Supporting inserting, loading, building index, search from index, GetVectorByIds for now.  Advanced features such as range search not yet implemented.

DO NOT MERGE this PR. I'll split this PR into multiple smaller PRs for code review. This PR is used for reviewers to grab a whole picture when reviewing individual PRs.

### CODE REVIEW

**Comment by ashkrisk** [comment]
Okay, so I figured out one of the things I was doing wrong. I needed to update `go.mod` and `go.sum` with the changes from https://github.com/milvus-io/milvus/pull/30400 as mentioned in https://github.com/milvus-io/milvus/issues/28381#issuecomment-1920758364. I originally thought this wouldn't affect the C++ build at all, but it looks like the version specified in `go.mod` is used to pull the proto files and generate code for the C++ build as well. Cherry picking that commit fixed the issue.

Now the C++ build is successful, but the Go build fails:
```
Building Milvus ...
# github.com/milvus-io/milvus/internal/util/indexcgowrapper
internal/util/indexcgowrapper/index.go:184:82: not enough arguments in call to (_Cfunc_BuildSparseFloatVecIndex)
  have (_Ctype_CIndex, _Ctype_long, *_Ctype_uchar)
  want (_Ctype_CIndex, _Ctype_long, _Ctype_long, unsafe.Pointer)
make: *** [Makefile:64: milvus] Error 1
```

Is there anything else that I'm missing?

**Comment by zhengbuqian** [comment]
hi @ashkrisk can you try again with the newest code? cpu-e2e test has just passed so I think it'll work. 

BTW you can try directly using the docker image built by the cpu-e2e process without worrying about the building process. The image tag I found in the [test log](https://jenkins.milvus.io:18080/blue/organizations/jenkins/Milvus%20HA%20CI/detail/PR-29421/11/pipeline/26) is `harbor.milvus.io/milvus/milvus:PR-29421-20240218-af65f4d51`.

**Comment by ashkrisk** [comment]
Thanks @zhengbuqian! I'm able to build successfully now. I can also run standard dense vector operations on the built image.

I still have some trouble with the sparse vector search though - `hello_sparse.py` doesn't run cleanly. Not sure if the issue is with `milvus` or `pymilvus` - insert fails with the following error:
```
pymilvus.exceptions.MilvusException: <MilvusException: (code=65535, message=nil indices in SparseFloatRow)>
```
Even though the inserted sparse matrix looks fine.

**Comment by zhengbuqian** [comment]
hi @ashkrisk I think that is caused by it not allowing empty sparse vector(0 at all dimensions). try update the vec dim and density to make it more dense to prevent `rand(num_entities, dim, density=density, format='csr')` from generating empty row(or at lease reduce the probability).

**Comment by ashkrisk** [comment]
Looks like that was the issue - all the matrices which failed to insert had at least one row with no elements. Everything looks fine as long as there are no zero vectors in the input!

**Comment by zhengbuqian** [comment]
Great! Let me know if you encountered any more issues so I can fix them!

**Comment by xiaofan-luan** [comment]
maybe we split this pr to smaller chunks for ease of review.

**Comment by xiaofan-luan** [comment]
I'm working on going through the pr today

**Comment by zhengbuqian** [comment]
@xiaofan-luan I have already splitted this PR:

1. #30357 segcore basics and index buidling
2. #30629 segcore search and get vector by id
3. #30630 all golang code
4. another PR to add more integration tests incoming

Due to the lack of stacking PR support in Github, each of those PRs also includes commits from all previous PRs, so from within each PR reviewer should only look at the last commit. @czs007 is also reviewing those PRs now.

File: pkg/util/typeutil/schema.go [hunk 1]
```diff
@@ -159,6 +159,12 @@ func estimateSizeBy(schema *schemapb.CollectionSchema, policy getVariableFieldLe
 					break
 				}
 			}
+		case schemapb.DataType_SparseFloatVector:
+			// TODO(SPARSE, zhengbuqian): size of sparse flaot vector
+			// varies depending on the number of non-zeros. Using sparse vector
+			// generated by SPLADE as reference and returning size of a sparse
+			// vector with 150 non-zeros.
+			res += 1200
```
Comment (xiaofan-luan) [inline pkg/util/typeutil/schema.go:167]: This is dangerous. 
We should not depends on any of this kind of estimation but so far lets keep it.
make this configurable

File: internal/storage/payload_reader.go [hunk 2]
```diff
@@ -413,6 +418,51 @@ func (r *PayloadReader) GetFloatVectorFromPayload() ([]float32, int, error) {
...
+	fieldData := &SparseFloatVectorFieldData{}
+
+	for _, value := range values {
+		if len(value)%8 != 0 {
+			return nil, -1, fmt.Errorf("invalid bytesData length")
+		}
+		elementCount := len(value) / 8
+		indices := make([]int32, elementCount)
+		values := make([]float32, elementCount)
+		for i := 0; i < elementCount; i++ {
```
Comment (xiaofan-luan) [inline internal/storage/payload_reader.go:443]: is there a reason we don't use Proto.marshal(SparseFloatRow) but serialize by customized code?

File: internal/core/src/common/Utils.h [hunk 3]
```diff
@@ -213,4 +214,67 @@ GetCommonPrefix(const std::string& str1, const std::string& str2) {
...
+
+inline int64_t
+SparseRowSerializedSize(const knowhere::sparse::SparseRow<float>& row) {
+    return row.size() * knowhere::sparse::SparseRow<float>::element_size();
+}
+
+// serialize a sparse row to bytes that will be written into binlog file.
+inline std::vector<uint8_t>
+SerializeSparseRow(const knowhere::sparse::SparseRow<float>& row) {
+    std::vector<uint8_t> buffer(SparseRowSerializedSize(row));
```
Comment (xiaofan-luan) [inline internal/core/src/common/Utils.h:266]: For all in memory format, let's use knowhere::sparse::SparseRow

Other that, for all Serdes, let's use protobuf

File: internal/core/src/segcore/ConcurrentVector.h [hunk 4]
```diff
@@ -162,27 +163,16 @@ class ConcurrentVectorImpl : public VectorBase {
...
-    }
-
-    void
-    grow_on_demand(int64_t element_count) {
-        auto chunk_count = upper_div(element_count, size_per_chunk_);
-        chunks_.emplace_to_at_least(chunk_count, Dim * element_count);
+    explicit ConcurrentVectorImpl(ssize_t elements_per_row,
+                                  int64_t size_per_chunk)
+        : VectorBase(size_per_chunk),
+          elements_per_row_(is_type_entire_row ? 1 : elements_per_row) {
```
[the remark this answers is not in the document]
Comment (zhengbuqian) [reply internal/core/src/segcore/ConcurrentVector.h:169]: `elements_per_row` is to denote how many `Type` objects we need to represent a single row:

In the old code:

* for a `float` scalar field, `Type = float, is_scalar = True`: each row value has 1 float
* for a dense `float` vector field, `Type = float, is_scalar = False`: each row value has `dim` floats

it implicitly assumes all scalar fields have a dim of 1, and all vector fields have a fixed dim. this assumption does not hold for sparse vector.

thus I renamed `is_scalar` to `is_type_entire_row`: if true, each row value consists of 1 `Type` object; or else each row value consists of `elements_per_row` of `Type` objects.

* for a `float` scalar field, `Type = float, is_type_entire_row = True`: each row value has `elements_per_row  =  1` float
* for a dense `float` vector field, `Type = float, is_type_entire_row = False`: each row has `elements_per_row = dim` floats
* for a sparse `float` vector field, `Type = SparseRow, is_type_entire_row = True`: each row has `elements_per_row  = 1` SparseRow

and we no longer have the notation of `dim` in the base class `ConcurrentVectorImpl`

File: internal/core/src/common/Utils.h (hunk 3 above)
Comment (zhengbuqian) [reply internal/core/src/common/Utils.h:266]: discussed offline: we'll use densely packed `idx, val, idx, val, ...` byte array everywhere: in `schema.proto::SparseFloatArray`, in golang/c++ memory, in persisted binlog, in placeholder of search requests, etc.

places we expect to see different format:

* in SDK and import utils we accept various formats as insert/search input, but we convert those before sending to milvus
* in restful endpoints we accept string representation like `{"indices": [102, 16, 18, ...], "values": [0.21, 0.11, 0.15, ...]}`

Comment (xiaofan-luan) [reply internal/core/src/common/Utils.h:266]: the design lgtm

File: internal/core/src/common/Schema.h [hunk 5]
```diff
@@ -132,9 +132,10 @@ class Schema {
         return fields_.at(field_id);
     }
 
+    // returned size of json, array and sparse float vector is estimated.
     auto
-    get_total_sizeof() const {
-        return total_sizeof_;
+    get_estimated_total_sizeof() const {
```
[the remark this answers is not in the document]
Comment (zhengbuqian) [reply internal/core/src/common/Schema.h:139]: added todo in code. do we have an issue tracking this? if not I can create one.

[the remark this answers is not in the document]
Comment (zhengbuqian) [reply internal/core/src/common/Schema.h:139]: ack, removing it.

File: internal/core/src/index/VectorMemIndex.cpp [hunk 6]
```diff
@@ -480,37 +480,66 @@ VectorMemIndex<T>::Build(const Config& config) {
...
+                    ->Dim());
+        }
+        std::vector<knowhere::sparse::SparseRow<float>> vec(total_rows);
+        int64_t offset = 0;
+        for (auto field_data : field_datas) {
+            auto ptr = static_cast<const knowhere::sparse::SparseRow<float>*>(
+                field_data->Data());
+            AssertInfo(ptr, "failed to cast field data to sparse rows");
+            for (size_t i = 0; i < field_data->Length(); ++i) {
+                // TODO(SPARSE): do not copy, add a method to force field_data
```
[the remark this answers is not in the document]
Comment (zhengbuqian) [reply internal/core/src/index/VectorMemIndex.cpp]: done

**Comment by xiaofan-luan** [comment]
@zhengbuqian is this ready to review and merge?

**Comment by zhengbuqian** [comment]
closing this PR. codes in this PR have been merged in smaller PRs, see linked issue.

**Pull request closed without merging**